File: client/wire_codec.c
/*
 * wire_codec.c
 * Purpose: Implementation of manual marshalling/unmarshalling for C client.
 * Design notes:
 * - Uses htons/htonl/ntohs/ntohl for 16/32-bit conversions.
 * - For 64-bit values, manually split into two 32-bit halves (high/low) and convert each.
 * - All writes advance the buffer pointer; caller manages buffer size.
 */

#include "wire_codec.h"
#include <string.h>

/* Platform-specific network byte order headers */
#ifdef _WIN32
    #include <winsock2.h>
#else
    #include <arpa/inet.h>  /* htons, htonl, ntohs, ntohl */
#endif

/* Write header: 16 bytes total */
int write_header(uint8_t *buf, const Header *h) {
    int offset = 0;                                      /* current write position */
    offset += write_u16(buf + offset, h->version);       /* write version uint16 */
    offset += write_u16(buf + offset, h->opCode);        /* write opCode uint16 */
    offset += write_u32(buf + offset, h->requestId);     /* write requestId uint32 */
    offset += write_u32(buf + offset, h->flags);         /* write flags uint32 */
    offset += write_u32(buf + offset, h->payloadLen);    /* write payloadLen uint32 */
    return offset;                                       /* should be 16 */
}

/* Read header: 16 bytes total */
int read_header(const uint8_t *buf, Header *h) {
    int offset = 0;                                      /* current read position */
    offset += read_u16(buf + offset, &h->version);       /* read version */
    offset += read_u16(buf + offset, &h->opCode);        /* read opCode */
    offset += read_u32(buf + offset, &h->requestId);     /* read requestId */
    offset += read_u32(buf + offset, &h->flags);         /* read flags */
    offset += read_u32(buf + offset, &h->payloadLen);    /* read payloadLen */
    return offset;                                       /* should be 16 */
}

/* Write uint16 in network byte order */
int write_u16(uint8_t *buf, uint16_t val) {
    uint16_t net = htons(val);                           /* convert to big-endian */
    memcpy(buf, &net, 2);                                /* copy 2 bytes */
    return 2;                                            /* bytes written */
}

/* Write uint32 in network byte order */
int write_u32(uint8_t *buf, uint32_t val) {
    uint32_t net = htonl(val);                           /* convert to big-endian */
    memcpy(buf, &net, 4);                                /* copy 4 bytes */
    return 4;                                            /* bytes written */
}

/* Write int64 in network byte order (split into two uint32) */
int write_i64(uint8_t *buf, int64_t val) {
    uint32_t high = (uint32_t)((val >> 32) & 0xFFFFFFFFUL); /* upper 32 bits */
    uint32_t low  = (uint32_t)(val & 0xFFFFFFFFUL);         /* lower 32 bits */
    int offset = 0;                                      /* write position */
    offset += write_u32(buf + offset, high);             /* write high word first (BE) */
    offset += write_u32(buf + offset, low);              /* write low word */
    return offset;                                       /* 8 bytes */
}

/* Read uint16 from network byte order */
int read_u16(const uint8_t *buf, uint16_t *val) {
    uint16_t net;                                        /* network byte order holder */
    memcpy(&net, buf, 2);                                /* copy 2 bytes */
    *val = ntohs(net);                                   /* convert to host order */
    return 2;                                            /* bytes read */
}

/* Read uint32 from network byte order */
int read_u32(const uint8_t *buf, uint32_t *val) {
    uint32_t net;                                        /* network byte order holder */
    memcpy(&net, buf, 4);                                /* copy 4 bytes */
    *val = ntohl(net);                                   /* convert to host order */
    return 4;                                            /* bytes read */
}

/* Read int64 from network byte order (two uint32) */
int read_i64(const uint8_t *buf, int64_t *val) {
    uint32_t high, low;                                  /* high and low words */
    int offset = 0;                                      /* read position */
    offset += read_u32(buf + offset, &high);             /* read high word */
    offset += read_u32(buf + offset, &low);              /* read low word */
    *val = ((int64_t)high << 32) | (int64_t)low;         /* combine into 64-bit */
    return offset;                                       /* 8 bytes */
}

/* Write length-prefixed string */
int write_string(uint8_t *buf, const char *str) {
    size_t len = strlen(str);                            /* compute string length */
    if (len > 0xFFFF) len = 0xFFFF;                      /* clamp to uint16 max */
    int offset = 0;                                      /* write position */
    offset += write_u16(buf + offset, (uint16_t)len);    /* write length prefix */
    memcpy(buf + offset, str, len);                      /* copy string bytes */
    offset += len;                                       /* advance by string length */
    return offset;                                       /* total bytes written */
}

/* Read length-prefixed string */
int read_string(const uint8_t *buf, char *out, size_t max_len) {
    uint16_t len;                                        /* string length from wire */
    int offset = 0;                                      /* read position */
    offset += read_u16(buf + offset, &len);              /* read length prefix */
    if (len >= max_len) return -1;                       /* buffer too small */
    memcpy(out, buf + offset, len);                      /* copy string bytes */
    out[len] = '\0';                                     /* null-terminate */
    offset += len;                                       /* advance by string length */
    return offset;                                       /* total bytes read */
}

File: common/WireCodec.java
/*
 * WireCodec.java
 * Purpose: Manual marshalling/unmarshalling helpers for the UDP protocol.
 * Design notes:
 * - All integers are encoded big-endian (network byte order).
 * - Header is fixed 16 bytes; helper methods write/read header and primitives.
 * - Strings are length-prefixed with uint16 length and UTF-8 bytes.
 * - Timestamps are 64-bit epochMillis (Java long) encoded big-endian.
 */

import java.nio.*;
import java.nio.charset.StandardCharsets;
import java.util.*;

public final class WireCodec {

    /*
     * Header structure used when building or parsing a UDP message.
     */
    public static final class Header {
        public int version;     // uint16 on wire
        public int opCode;       // uint16 on wire
        public long requestId;   // uint32 on wire (stored as long to avoid sign)
        public long flags;       // uint32 on wire (stored as long)
        public int payloadLen;   // uint32 on wire, but fits in Java int
    }

    // Allocate a ByteBuffer with big-endian order for writing messages
    public static ByteBuffer allocate(int capacity) {
        return ByteBuffer.allocate(capacity).order(Protocol.BYTE_ORDER); // ensure BE
    }

    // Wrap an existing byte array for reading with big-endian order
    public static ByteBuffer wrap(byte[] data) {
        return ByteBuffer.wrap(data).order(Protocol.BYTE_ORDER); // ensure BE
    }

    // Write header at position 0..15
    public static void writeHeader(ByteBuffer buf, Header h) {
        buf.putShort((short) (h.version & 0xFFFF));                // version uint16
        buf.putShort((short) (h.opCode & 0xFFFF));                 // opCode uint16
        buf.putInt((int) (h.requestId & 0xFFFFFFFFL));            // requestId uint32
        buf.putInt((int) (h.flags & 0xFFFFFFFFL));                // flags uint32
        buf.putInt(h.payloadLen);                                  // payload length uint32
    }

    // Read header from buffer starting at current position
    public static Header readHeader(ByteBuffer buf) {
        Header h = new Header();                                   // allocate header
        h.version = Short.toUnsignedInt(buf.getShort());           // read uint16 version
        h.opCode = Short.toUnsignedInt(buf.getShort());            // read uint16 opCode
        h.requestId = Integer.toUnsignedLong(buf.getInt());        // read uint32 requestId
        h.flags = Integer.toUnsignedLong(buf.getInt());            // read uint32 flags
        h.payloadLen = buf.getInt();                               // read uint32 payload length
        return h;                                                  // return parsed header
    }

    // Write a uint16 value
    public static void writeU16(ByteBuffer buf, int v) {
        buf.putShort((short) (v & 0xFFFF));                        // store as 16-bit
    }

    // Write a uint32 value
    public static void writeU32(ByteBuffer buf, long v) {
        buf.putInt((int) (v & 0xFFFFFFFFL));                       // store low 32 bits
    }

    // Write a uint64/signed long (used for timestamps and booking id)
    public static void writeI64(ByteBuffer buf, long v) {
        buf.putLong(v);                                            // store 64-bit big-endian
    }

    // Read helpers
    public static int readU16(ByteBuffer buf) { return Short.toUnsignedInt(buf.getShort()); }
    public static long readU32(ByteBuffer buf) { return Integer.toUnsignedLong(buf.getInt()); }
    public static long readI64(ByteBuffer buf) { return buf.getLong(); }

    // Write a length-prefixed UTF-8 string
    public static void writeString(ByteBuffer buf, String s) {
        byte[] bytes = s.getBytes(StandardCharsets.UTF_8);         // encode UTF-8 bytes
        if (bytes.length > 0xFFFF) throw new IllegalArgumentException("string too long"); // bound check
        writeU16(buf, bytes.length);                               // prefix with uint16 length
        buf.put(bytes);                                            // write bytes
    }

    // Read a length-prefixed UTF-8 string
    public static String readString(ByteBuffer buf) {
        int len = readU16(buf);                                    // read uint16 length
        byte[] bytes = new byte[len];                              // allocate
        buf.get(bytes);                                            // read bytes
        return new String(bytes, StandardCharsets.UTF_8);          // decode to String
    }

    // Compute total message buffer: header + payload size
    public static ByteBuffer newMessageBuffer(int payloadLength) {
        return allocate(Protocol.HEADER_LEN + payloadLength);      // allocate total buffer
    }

    private WireCodec() { /* utility */ }
}
